### PR TITLE
Fix logic in compareTypesForCast

### PR DESCRIPTION
**Pull request by AndyAyersMS**
We weren't careful enough with `__Canon` in some cases, which lead to unsafely
returning `MustNot` when the cast outcome was not determined at jit time.

Add an extra check, update comments, and add some test cases.

Addresses the failures seen in #1195 (which was reverted).

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
cc @jkotas @EgorBo 

No diffs seen in (x64 windows pmi Fx); current callers of `checkTypesForCast` not impacted.

**Comment by EgorBo** [comment]
NOTE: it always gives up if `toHnd` is `Nullable<>` here https://github.com/dotnet/runtime/blob/master/src/coreclr/src/vm/jitinterface.cpp#L4474-L4477

probably makes sense to return `MustNot` if `fromHnd` is not value type.

File: src/coreclr/src/vm/jitinterface.cpp [hunk 1]
```diff
@@ -4495,16 +4495,29 @@ TypeCompareState CEEInfo::compareTypesForCast(
             {
                 result = TypeCompareState::Must;
             }
-            // For negative results, the unknown type parameter in
-            // fromClass might match some instantiated interface,
-            // either directly or via variance.
+            // We have __Canon parameter(s) in fromClass, somehwere.
```
Comment (jkotas) [inline src/coreclr/src/vm/jitinterface.cpp]:
```suggestion
            // We have __Canon parameter(s) in fromClass, somewhere.
```

File: src/coreclr/src/tools/Common/JitInterface/CorInfoImpl.cs [hunk 2]
```diff
@@ -1757,16 +1757,29 @@ private TypeCompareState compareTypesForCast(CORINFO_CLASS_STRUCT_* fromClass, C
                     {
                         result = TypeCompareState.Must;
                     }
-                    // For negative results, the unknown type parameter in
-                    // fromClass might match some instantiated interface,
-                    // either directly or via variance.
+                    // We have __Canon parameter(s) in fromClass, somehwere.
```
Comment (jkotas) [inline src/coreclr/src/tools/Common/JitInterface/CorInfoImpl.cs]:
```suggestion
                    // We have __Canon parameter(s) in fromClass, somewhere.
```

**Review by jkotas — APPROVED** [review]
Thanks!

**Comment by AndyAyersMS** [comment]
Formatting failure seems bogus; I did not change the code in the jit, and the patch file produced is empty.

**Comment by AndyAyersMS** [comment]
x86 libraries "System.Threading.Tasks.Tests Work Item" has failed twice now, leaving no diagnostic output to work with. Fairly sure this is unrelated to this change and it fails intermittently in master too.

Looks like this could be #2271 -- both failures  in the above CI runs happen just after the 15 min mark.

image

**Pull request merged**